File: libraries/ios/Sources/VoidhashCore/Network/CircuitBreaker.swift
import Foundation

/// State of a ``CircuitBreaker`` host entry.
public enum CircuitBreakerState: String, Sendable, Equatable {
    /// Requests flow normally.
    case closed
    /// Requests are refused; the SDK serves cached state instead.
    case open
    /// One probe request is allowed through to test whether the host recovered.
    case halfOpen
}

/// Permission to issue one request, which must be handed back through
/// ``CircuitBreaker/release(_:retryableFailure:)``.
///
/// A half-open breaker grants exactly one of these, so losing one would wedge the breaker shut
/// forever. Releasing is therefore the caller's obligation on *every* exit, including the paths
/// where the work was never started.
public struct CircuitBreakerPermit: Sendable, Equatable {
    let host: String
    let identifier: UInt64
    /// Whether this permit is the single probe of a half-open breaker.
    public let isProbe: Bool
}

/// Per-host breaker guarding the SDK against hammering an origin that is down.
///
/// Opens after ``failureThreshold`` consecutive retryable failures. The first cooldown is 30 s
/// and doubles on every re-open up to 5 minutes. Foreground and connectivity-restored events
/// half-open the breaker rather than resetting it, so a genuinely dead host still gets exactly
/// one probe. Authentication failures and other 4xx never count.
public actor CircuitBreaker {
    /// Consecutive retryable failures that open the breaker.
    public static let failureThreshold = 5
    /// First cooldown before a probe is allowed through.
    public static let initialCooldownMilliseconds: Double = 30_000
    /// Ceiling the doubling cooldown stops at.
    public static let maxCooldownMilliseconds: Double = 300_000
    /// How long an outstanding probe is honoured before it is treated as abandoned.
    ///
    /// A caller that is cancelled, deallocated or hung mid-probe would otherwise hold the only
    /// slot for the lifetime of the process; the timeout matches the request timeout, so a probe
    /// that could still be in flight is never stolen.
    public static let probeTimeoutMilliseconds: Double =
        NetworkPolicy.requestTimeoutSeconds * 1000

    private struct HostState {
        var consecutiveFailures = 0
        var state: CircuitBreakerState = .closed
        var cooldownMilliseconds = CircuitBreaker.initialCooldownMilliseconds
        var openedAt: Double = 0
        var probeIdentifier: UInt64?
        var probeStartedAt: Double = 0
    }

    private let clock: any VoidhashClock
    private let diagnostics: DiagnosticEmitter
    private var hosts: [String: HostState] = [:]
    private var nextIdentifier: UInt64 = 1

    public init(clock: any VoidhashClock, diagnostics: DiagnosticEmitter = DiagnosticEmitter(nil)) {
        self.clock = clock
        self.diagnostics = diagnostics
    }

    /// Whether a request to `host` would be admitted, without consuming anything.
    ///
    /// Use it to decide whether a read is worth attempting; ``acquire(host:)`` is what actually
    /// admits it.
    public func shouldAttempt(host: String) -> Bool {
        guard let entry = hosts[host] else {
            return true
        }
        switch entry.state {
        case .closed:
            return true
        case .open:
            return clock.now() - entry.openedAt >= entry.cooldownMilliseconds
        case .halfOpen:
            return true
        }
    }

    /// Admits one request, or returns `nil` when the breaker is holding the host shut.
    public func acquire(host: String) -> CircuitBreakerPermit? {
        var entry = hosts[host] ?? HostState()
        let timestamp = clock.now()

        if entry.state == .open {
            guard timestamp - entry.openedAt >= entry.cooldownMilliseconds else {
                hosts[host] = entry
                return nil
            }
            entry.state = .halfOpen
            entry.probeIdentifier = nil
        }

        if entry.state == .halfOpen {
            // An outstanding probe that has outlived the request timeout is treated as lost,
            // which is what keeps a cancelled caller from wedging the breaker permanently.
            if entry.probeIdentifier != nil,
                timestamp - entry.probeStartedAt < CircuitBreaker.probeTimeoutMilliseconds
            {
                hosts[host] = entry
                return nil
            }
            let identifier = takeIdentifier()
            entry.probeIdentifier = identifier
            entry.probeStartedAt = timestamp
            hosts[host] = entry
            return CircuitBreakerPermit(host: host, identifier: identifier, isProbe: true)
        }

        hosts[host] = entry
        return CircuitBreakerPermit(host: host, identifier: takeIdentifier(), isProbe: false)
    }

    /// Hands a permit back.
    ///
    /// - Parameter retryableFailure: `nil` when the request succeeded, `true` for a transport or
    ///   retryable-status failure, `false` for a client error the host is not to blame for.
    public func release(_ permit: CircuitBreakerPermit, retryableFailure: Bool?) {
        guard var entry = hosts[permit.host] else {
            return
        }
        // Only the permit that still owns the probe slot may clear it: a late release from an
        // abandoned probe must not free the slot the replacement probe is using.
        let ownsProbe = permit.isProbe && entry.probeIdentifier == permit.identifier
        if ownsProbe {
            entry.probeIdentifier = nil
        }

        guard let retryableFailure else {
            let wasOpen = entry.state != .closed
            entry.consecutiveFailures = 0
            entry.state = .closed
            entry.cooldownMilliseconds = CircuitBreaker.initialCooldownMilliseconds
            entry.probeIdentifier = nil
            hosts[permit.host] = entry
            if wasOpen {
                diagnostics.emit(
                    .breaker, code: "CIRCUIT_CLOSED", operation: "network.breaker",
                    message: "Requests to \(permit.host) resumed")
            }
            return
        }

        guard retryableFailure else {
            // A client error says nothing about the host's health, so it neither counts towards
            // opening the breaker nor closes it.
            hosts[permit.host] = entry
            return
        }

        entry.consecutiveFailures += 1

        if ownsProbe || entry.state == .halfOpen {
            entry.cooldownMilliseconds = min(
                entry.cooldownMilliseconds * 2, CircuitBreaker.maxCooldownMilliseconds)
            open(&entry, host: permit.host, reason: "the probe failed")
            hosts[permit.host] = entry
            return
        }

        if entry.state == .closed, entry.consecutiveFailures >= CircuitBreaker.failureThreshold {
            open(
                &entry, host: permit.host,
                reason: "\(entry.consecutiveFailures) consecutive failures")
        }
        hosts[permit.host] = entry
    }

    /// Releases a permit for work that never ran, without judging the host.
    public func abandon(_ permit: CircuitBreakerPermit) {
        release(permit, retryableFailure: false)
    }

    /// Reported state of `host`.
    public func state(host: String) -> CircuitBreakerState {
        return hosts[host]?.state ?? .closed
    }

    /// Moves every non-closed breaker to half-open so the next request probes immediately.
    ///
    /// Called on app foreground and on connectivity-restored. Does not reset the cooldown ladder.
    /// A probe that has outlived ``probeTimeoutMilliseconds`` is treated as lost so a wedged
    /// half-open host gets another chance; one still inside the timeout keeps its slot, so a
    /// foreground during an in-flight probe cannot hand out a second one.
    public func halfOpenAll() {
        let timestamp = clock.now()
        for (host, var entry) in hosts where entry.state != .closed {
            entry.state = .halfOpen
            if entry.probeIdentifier != nil,
                timestamp - entry.probeStartedAt >= CircuitBreaker.probeTimeoutMilliseconds
            {
                entry.probeIdentifier = nil
            }
            hosts[host] = entry
        }
    }

    /// Drops all breaker state.
    public func reset() {
        hosts = [:]
    }

    // One diagnostic per transition into `.open`: a breaker that flaps would otherwise emit on
    // every refused request.
    private func open(_ entry: inout HostState, host: String, reason: String) {
        let wasOpen = entry.state == .open
        entry.state = .open
        entry.openedAt = clock.now()
        entry.probeIdentifier = nil
        guard !wasOpen else {
            return
        }
        diagnostics.emit(
            .breaker, code: "CIRCUIT_OPEN", operation: "network.breaker", retryable: true,
            message:
                "Requests to \(host) paused after \(reason); retrying in \(Int(entry.cooldownMilliseconds)) ms"
        )
    }

    private func takeIdentifier() -> UInt64 {
        defer { nextIdentifier += 1 }
        return nextIdentifier
    }
}

File: libraries/android/sdk/src/main/java/com/voidhash/sdk/network/CircuitBreaker.kt
package com.voidhash.sdk.network

import com.voidhash.sdk.diagnostics.DiagnosticEmitter
import com.voidhash.sdk.diagnostics.VoidhashDiagnosticKind
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicBoolean

/** How many consecutive retryable failures open a host's breaker. */
private const val FAILURE_THRESHOLD = 5

/** First open window; each consecutive re-open doubles it. */
private const val INITIAL_OPEN_MS = 30_000L

/** Ceiling of the doubling open window. */
private const val MAX_OPEN_MS = 5 * 60_000L

/**
 * How long a half-open probe may hold its slot.
 *
 * Matches the HTTP call timeout: a probe that has not reported by then is not coming back,
 * and leaving the slot claimed would wedge the host in [CircuitState.HALF_OPEN] forever.
 */
internal const val PROBE_TIMEOUT_MS = 30_000L

/** Observable state of one host's breaker. */
enum class CircuitState {
    /** Requests pass through. */
    CLOSED,

    /** Requests are refused; the cache answers instead. */
    OPEN,

    /** One probe request is allowed through to test the host. */
    HALF_OPEN,
}

/**
 * A granted permission to make one request.
 *
 * Every acquired permit must be handed back exactly once — through
 * [CircuitBreaker.recordSuccess], [CircuitBreaker.recordFailure], or [CircuitBreaker.release]
 * from a `finally` — or a half-open host keeps the probe slot claimed and never recovers.
 * Handing the same permit back twice is a no-op.
 */
class CircuitPermit internal constructor(
    internal val host: String,
    internal val isProbe: Boolean,
    internal val acquiredAt: Long,
) {
    internal val settled = AtomicBoolean(false)

    internal fun claim(): Boolean = settled.compareAndSet(false, true)
}

/**
 * Per-host circuit breaker.
 *
 * A host that fails five retryable attempts in a row is skipped entirely for 30 s, so an
 * outage costs one timeout rather than one per call. The window doubles on every
 * consecutive re-open up to five minutes. App foreground and connectivity-restored
 * half-open the breaker rather than resetting it, so recovery still costs one probe.
 *
 * Authentication failures and other 4xx verdicts never count: they are answers, not outages.
 */
class CircuitBreaker(
    private val clock: SdkClock = SystemSdkClock,
    private val diagnostics: DiagnosticEmitter = DiagnosticEmitter(),
    private val probeTimeoutMs: Long = PROBE_TIMEOUT_MS,
) {
    private class HostState {
        var consecutiveFailures: Int = 0
        var openUntil: Long = 0L
        var openWindowMs: Long = INITIAL_OPEN_MS
        var probeStartedAt: Long? = null
        var reportedOpen: Boolean = false
    }

    private val hosts = ConcurrentHashMap<String, HostState>()

    /** The breaker state [host] is currently in. */
    fun state(host: String): CircuitState {
        val entry = hosts[host] ?: return CircuitState.CLOSED
        return synchronized(entry) { stateLocked(entry) }
    }

    /**
     * Asks for permission to reach [host], returning `null` when the circuit refuses.
     *
     * In [CircuitState.HALF_OPEN] exactly one caller is granted the probe; the rest are
     * refused until that probe is settled or its timeout elapses.
     */
    fun acquire(host: String, operation: String = "network"): CircuitPermit? {
        val entry = hosts.getOrPut(host) { HostState() }
        return synchronized(entry) {
            when (stateLocked(entry)) {
                CircuitState.CLOSED -> CircuitPermit(host, isProbe = false, acquiredAt = clock.now())
                CircuitState.HALF_OPEN -> {
                    entry.probeStartedAt = clock.now()
                    CircuitPermit(host, isProbe = true, acquiredAt = clock.now())
                }
                CircuitState.OPEN -> {
                    if (!entry.reportedOpen) {
                        entry.reportedOpen = true
                        diagnostics.emit(
                            VoidhashDiagnosticKind.BREAKER,
                            code = "CIRCUIT_OPEN",
                            operation = operation,
                            retryable = true,
                            message = "Skipping $host: the circuit is open after " +
                                "${entry.consecutiveFailures} consecutive failures",
                        )
                    }
                    null
                }
            }
        }
    }

    /** Records a completed request; closes the breaker and clears the doubling window. */
    fun recordSuccess(permit: CircuitPermit?) {
        val entry = permit?.let { hosts[it.host] } ?: return
        if (!permit.claim()) return
        synchronized(entry) {
            entry.consecutiveFailures = 0
            entry.openUntil = 0L
            entry.openWindowMs = INITIAL_OPEN_MS
            entry.probeStartedAt = null
            entry.reportedOpen = false
        }
    }

    /** Records a retryable transport failure; opens the breaker once the threshold is reached. */
    fun recordFailure(permit: CircuitPermit?, operation: String = "network") {
        val entry = permit?.let { hosts[it.host] } ?: return
        if (!permit.claim()) return

        val opened = synchronized(entry) {
            entry.probeStartedAt = null
            entry.consecutiveFailures += 1

            // A failed probe re-opens immediately on the doubled window: the host answered
            // the one question the half-open state asked.
            if (permit.isProbe) {
                entry.openWindowMs = minOf(entry.openWindowMs * 2, MAX_OPEN_MS)
                entry.openUntil = clock.now() + entry.openWindowMs
                entry.reportedOpen = false
                return@synchronized true
            }

            if (entry.consecutiveFailures >= FAILURE_THRESHOLD && entry.openUntil <= clock.now()) {
                entry.openUntil = clock.now() + entry.openWindowMs
                entry.reportedOpen = false
                true
            } else {
                false
            }
        }

        if (opened) {
            diagnostics.emit(
                VoidhashDiagnosticKind.BREAKER,
                code = "CIRCUIT_OPENED",
                operation = operation,
                retryable = true,
                message = "Opened the circuit for ${permit.host}",
            )
        }
    }

    /**
     * Hands an unreported permit back without a verdict.
     *
     * Cancellation, a bug in the request path, or any throw that is not a transport failure
     * ends here. Nothing is counted — the attempt proved nothing about the host — but the
     * probe slot is freed, which is the whole point of calling this from a `finally`.
     */
    fun release(permit: CircuitPermit?) {
        val entry = permit?.let { hosts[it.host] } ?: return
        if (!permit.claim()) return
        if (!permit.isProbe) return
        synchronized(entry) { entry.probeStartedAt = null }
    }

    /**
     * Moves every open host to [CircuitState.HALF_OPEN].
     *
     * Called on app foreground and on connectivity restored: both are strong evidence the
     * network changed, but neither proves the host recovered, so the next request is still
     * a single probe. A probe that is genuinely still in flight keeps its slot; only one
     * that has outlived [probeTimeoutMs] — and therefore is never reporting back — is
     * cleared, so a wedged host can recover here too.
     */
    fun halfOpenAll() {
        for (entry in hosts.values) {
            synchronized(entry) {
                if (entry.openUntil > clock.now()) {
                    entry.openUntil = 0L
                    entry.reportedOpen = false
                }
                val probeStartedAt = entry.probeStartedAt
                if (probeStartedAt != null && clock.now() - probeStartedAt >= probeTimeoutMs) {
                    entry.probeStartedAt = null
                    entry.reportedOpen = false
                }
            }
        }
    }

    private fun stateLocked(entry: HostState): CircuitState = when {
        entry.consecutiveFailures < FAILURE_THRESHOLD -> CircuitState.CLOSED
        entry.openUntil > clock.now() -> CircuitState.OPEN
        // A probe whose timeout has elapsed is gone; its slot goes back to the next caller.
        entry.probeStartedAt?.let { clock.now() - it < probeTimeoutMs } == true -> CircuitState.OPEN
        else -> CircuitState.HALF_OPEN
    }
}
